### PR TITLE
Don't use a separate call context for subchannel calls

### LINKED ISSUE #15927: Subchannel call context not the same in filter initialization and batch starting
When a filter is registered on the subchannel, the context that is passed to init_call_elem() through args->context is the subchannel context. However, the context that is passed to start_transport_stream_op_batch through batch->payload->context is the call context. This happens in the non-retry case (common case), but doesn't appear to happen in the retry case (uncommon).

Expected behavior: 
The context that is passed to a filter should be consistent across all of the filter's methods. 

There are a couple of proposed solutions:
1) Pass in subchannel context in all batches and propagate anything in parent context to subchannel context when we create it.
2) Find a different approach for what we’re using the subchannel context for

### PR DESCRIPTION
**Pull request by markdroth**
Fixes #15927.

This changes the grpclb code to pass the client stats object down into the subchannel call by attaching it as user data on the LB token metadata element.  This eliminates the need to store it in the subchannel call context, which means that we can stop using a separate call context in the subchannel call and instead just pass down the call context from the parent call.

Because we are no longer using the call context to pass anything down from the LB policy into the subchannel call, I have removed it from the LB policy API.  This is yet another small step toward having a public LB policy API.

Some notes:

- In order to make this work, I added support for attaching user data to `grpc_mdelem`s of storage type `GRPC_MDELEM_STORAGE_ALLOCATED`.  Previously, user data was only supported for type `GRPC_MDELEM_STORAGE_INTERNED`.
- In the xds policy, I have removed the code for passing the client stats object down into the subchannel call.  I suspect this is not needed at all in xds, but if it is, we can re-add it later, when we know how it will actually need to work.
- The subchannel call context will likely need more work when we eventually get around to implementing hedging support, but we can cross that bridge when we get to it.
- I have also changed client_auth_filter to set the security context at call stack initialization time, rather than waiting for the first batch of ops.  This was previously attempted in #15879 but didn't work due to the bug that this PR fixes.

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/18094)

### CODE REVIEW

**Comment by hcaseyal** [comment]
@markdroth Does this PR depend on the recently reverted lb_picker PR?

**Comment by AspirinSJL** [comment]
@hcaseyal I just had a quick glance. I think it doesn't depend on https://github.com/grpc/grpc/pull/17770. But since https://github.com/grpc/grpc/pull/17770 is a major change, there are some conflicts now after it's reverted.

**Comment by markdroth** [comment]
I could probably rebase this so that it could go in without the LB picker API change, but I'm not going to bother to do so.  I have a whole bunch of PRs depending on the LB picker API change, so getting that rolled forward is my top priority right now, and there's no urgency to push this PR forward before that happens.

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc [hunk 1]
```diff
@@ -562,15 +554,15 @@ GrpcLb::Picker::PickResult GrpcLb::Picker::Pick(PickState* pick,
...
-          DestroyClientStats;
+    GPR_ASSERT(!GRPC_MDISNULL(lb_token));
+    GPR_ASSERT(grpc_metadata_batch_add_tail(pick->initial_metadata,
+                                            &pick->lb_token_mdelem_storage,
+                                            GRPC_MDELEM_REF(lb_token))
+                    == GRPC_ERROR_NONE);
+    GrpcLbClientStats* client_stats = static_cast<GrpcLbClientStats*>(
+        grpc_mdelem_get_user_data(lb_token, GrpcLbClientStats::Destroy));
+    if (client_stats != nullptr) {
+      client_stats->AddCallStarted();
```
Comment (hcaseyal) [inline src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: Previously, we were calling this in the constructor of the call_data. Are there any implications to calling it here instead?

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc [hunk 2]
```diff
@@ -1534,7 +1526,8 @@ grpc_channel_args* GrpcLb::CreateRoundRobinPolicyArgsLocked() {
   ServerAddressList* addresses = &tmp_addresses;
   bool is_backend_from_grpclb_load_balancer = false;
   if (serverlist_ != nullptr) {
-    tmp_addresses = serverlist_->GetServerAddressList();
+    tmp_addresses = serverlist_->GetServerAddressList(
```
Comment (hcaseyal) [inline src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: Do we have tests for both branches of this code path?

File: src/core/lib/security/transport/client_auth_filter.cc [hunk 3]
```diff
@@ -41,12 +41,43 @@
 #define MAX_CREDENTIALS_METADATA_COUNT 4
 
 namespace {
+
+/* We can have a per-channel credentials. */
```
Comment (hcaseyal) [inline src/core/lib/security/transport/client_auth_filter.cc]: Would it be possible to add a test for the bug this PR fixes?

**Comment by markdroth** [comment]
@hcaseyal @AspirinSJL Gentle ping for re-review.  Thanks!

**Comment by markdroth** [comment]
Known issues: #17801

**Pull request merged**